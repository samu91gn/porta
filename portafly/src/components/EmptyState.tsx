import React from 'react'
import {
  Title,
  EmptyState,
  EmptyStateVariant,
  EmptyStateIcon,
  Bullseye,
  EmptyStateBody
} from '@patternfly/react-core'
import { CubesIcon, SearchIcon } from '@patternfly/react-icons'
import { IRow } from '@patternfly/react-table'

export interface ISimpleEmptyStateProps {
  msg: string
}

const SimpleEmptyState: React.FunctionComponent<ISimpleEmptyStateProps> = ({ msg }) => (
  <EmptyState variant={EmptyStateVariant.full}>
    <EmptyStateIcon icon={CubesIcon} />
    <Title headingLevel="h5" size="lg">
      {msg}
    </Title>
  </EmptyState>
)

type ICreateTableEmptyState = (title: string, body: string, button?: JSX.Element) => IRow[]

const CreateTableEmptyState: ICreateTableEmptyState = (title, body, button) => [{
  heightAuto: true,
  cells: [{
    props: { colSpan: 8 },
    title: (
      <Bullseye>
        <EmptyState variant={EmptyStateVariant.small}>
          <EmptyStateIcon icon={SearchIcon} />
          <Title headingLevel="h2" size="lg">
            {title}
          </Title>
          <EmptyStateBody>
            {body}
          </EmptyStateBody>
          {button}
        </EmptyState>
      </Bullseye>
    )
  }]
}]

export { SimpleEmptyState, CreateTableEmptyState }
