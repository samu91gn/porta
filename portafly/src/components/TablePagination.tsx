import React, { FunctionComponent, useMemo, useReducer } from 'react'
import { Pagination, OnSetPage, OnPerPageSelect } from '@patternfly/react-core'

interface IPagination {
  itemCount: number
  state: PaginationState
  dispatch: React.Dispatch<PaginationAction>
}

const perPageOptions = [
  { title: '5', value: 5 },
  { title: '20', value: 20 },
  { title: '50', value: 50 }
]

const TablePagination: FunctionComponent<IPagination> = ({
  itemCount,
  state,
  dispatch
}) => {
  const { perPage, page } = state

  const onSetPage: OnSetPage = (ev, newPage, _perPage, startIdx, endIdx) => {
    dispatch({
      type: 'setPage',
      payload: {
        page: newPage,
        startIdx,
        endIdx
      }
    })
  }

  const onPerPageSelect: OnPerPageSelect = (ev, newPerPage, newPage, startIdx, endIdx) => {
    dispatch({
      type: 'setPerPage',
      payload: {
        page: newPage,
        perPage: newPerPage,
        startIdx,
        endIdx
      }
    })
  }

  const isCompact = useMemo(() => (itemCount / perPage) < 3, [itemCount, perPage])

  return (
    <Pagination
      itemCount={itemCount}
      perPage={perPage}
      page={page}
      onSetPage={onSetPage}
      onPerPageSelect={onPerPageSelect}
      perPageOptions={perPageOptions}
      isCompact={isCompact}
    />
  )
}

type PaginationState = {
  page: number
  perPage: number
  startIdx?: number
  endIdx?: number
}

type PaginationAction = {
  type: 'setPage' | 'setPerPage'
  payload: Partial<PaginationState>
}

const paginationReducer = (state: PaginationState, action: PaginationAction):
PaginationState => {
  switch (action.type) {
    case 'setPage':
    case 'setPerPage':
      return { ...state, ...action.payload }
    default:
      return { ...state }
  }
}

const usePaginationReducer = () => useReducer(paginationReducer, {
  page: 0,
  perPage: perPageOptions[0].value,
  startIdx: 0,
  endIdx: perPageOptions[0].value
})

export { TablePagination, usePaginationReducer }
