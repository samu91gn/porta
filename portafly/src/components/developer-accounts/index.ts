export * from './ActionsDropdown'
export * from './BulkActionSelector'
export * from './SearchWidget'
export * from './DeveloperAccountsTable'
export * from './SendEmailModal'
export * from './ChangePlanModal'
export * from './ChangeStateModal'
