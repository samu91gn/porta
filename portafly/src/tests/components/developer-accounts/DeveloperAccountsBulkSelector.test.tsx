/* eslint-disable react/jsx-props-no-spreading */
import React from 'react'

import { render } from 'tests/custom-render'
import { DeveloperAccountsBulkSelector } from 'components'
import { RenderResult, fireEvent, within } from '@testing-library/react'

function renderWrapper(props?: any): RenderResult {
  return render(
    <DeveloperAccountsBulkSelector
      onSelectAll={jest.fn()}
      onSelectPage={jest.fn()}
      pageCount={10}
      allCount={25}
      {...props}
    />
  )
}

it('expands and collapses properly', () => {
  const wrapper = renderWrapper()
  expect(wrapper.queryByRole('menu')).not.toBeInTheDocument()

  const button = wrapper.getByRole('button')
  fireEvent.click(button)

  const menu = within(wrapper.getByRole('menu'))
  expect(menu.getByText('accounts_table.data_toolbar.bulk_selector.none')).toBeInTheDocument()
  expect(menu.getByText('accounts_table.data_toolbar.bulk_selector.page')).toBeInTheDocument()
  expect(menu.getByText('accounts_table.data_toolbar.bulk_selector.all')).toBeInTheDocument()

  fireEvent.click(button)
  expect(wrapper.queryByRole('menu')).not.toBeInTheDocument()
})

describe('when it is checked', () => {
  let wrapper: RenderResult

  beforeEach(() => {
    wrapper = renderWrapper({ isChecked: true })
  })

  it('should render properly', () => {
    expect(wrapper.container.firstChild).toMatchSnapshot()
  })
})

describe('when it is unchecked', () => {
  let wrapper: RenderResult

  beforeEach(() => {
    wrapper = renderWrapper({ isChecked: false })
  })

  it('should render properly', () => {
    expect(wrapper.container.firstChild).toMatchSnapshot()
  })
})
