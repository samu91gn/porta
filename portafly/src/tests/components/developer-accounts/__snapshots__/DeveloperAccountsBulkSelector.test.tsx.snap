// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly when all are selected 1`] = `
<div
  class="pf-c-dropdown"
  id="checkbox_dropdown_id"
>
  <div
    class="pf-c-dropdown__toggle pf-m-split-button"
  >
    <label
      class="pf-c-dropdown__toggle-check"
      for="developer-accounts-bulk-checkbox"
    >
      <input
        aria-invalid="false"
        aria-label="Select all"
        checked=""
        data-testid="developer-accounts-bulk-checkbox"
        id="developer-accounts-bulk-checkbox"
        type="checkbox"
      />
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Select"
      class="pf-c-dropdown__toggle-button"
      id="pf-toggle-id-3"
      type="button"
    >
      <span
        class="pf-c-dropdown__toggle-text"
      >
        accounts_table.data_toolbar.bulk_selector.label
      </span>
      <svg
        aria-hidden="true"
        class="pf-c-dropdown__toggle-icon"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 320 512"
        width="1em"
      >
        <path
          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          transform=""
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`should render properly when none are selected 1`] = `
<div
  class="pf-c-dropdown"
  id="checkbox_dropdown_id"
>
  <div
    class="pf-c-dropdown__toggle pf-m-split-button"
  >
    <label
      class="pf-c-dropdown__toggle-check"
      for="developer-accounts-bulk-checkbox"
    >
      <input
        aria-invalid="false"
        aria-label="Select all"
        data-testid="developer-accounts-bulk-checkbox"
        id="developer-accounts-bulk-checkbox"
        type="checkbox"
      />
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Select"
      class="pf-c-dropdown__toggle-button"
      id="pf-toggle-id-5"
      type="button"
    >
      
      <svg
        aria-hidden="true"
        class=""
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 320 512"
        width="1em"
      >
        <path
          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          transform=""
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`should render properly when some are selected 1`] = `
<div
  class="pf-c-dropdown"
  id="checkbox_dropdown_id"
>
  <div
    class="pf-c-dropdown__toggle pf-m-split-button"
  >
    <label
      class="pf-c-dropdown__toggle-check"
      for="developer-accounts-bulk-checkbox"
    >
      <input
        aria-invalid="false"
        aria-label="Select all"
        data-testid="developer-accounts-bulk-checkbox"
        id="developer-accounts-bulk-checkbox"
        type="checkbox"
      />
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Select"
      class="pf-c-dropdown__toggle-button"
      id="pf-toggle-id-4"
      type="button"
    >
      <span
        class="pf-c-dropdown__toggle-text"
      >
        accounts_table.data_toolbar.bulk_selector.label
      </span>
      <svg
        aria-hidden="true"
        class="pf-c-dropdown__toggle-icon"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 320 512"
        width="1em"
      >
        <path
          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          transform=""
        />
      </svg>
    </button>
  </div>
</div>
`;
